### PR TITLE
[4.2] [DRAFT]  Global Delete Trash - New Feature

### PR DESCRIPTION
**Pull request by brianteeman**
It has been requested several times that there was a single place where you could go and "empty the trash" for all components.

This new component is based on com_checkin

At this stage 
[ ] I am sure the code in the model can be improved
[ ] What should the default permissions be or should it be hard coded restricted to super users only like joomlaupdate is
[ ] To do - add a quickicon on the home dashboard ?

### testing
Apply patch and then discover the component
You will then find it in the Maintenance section of the system dashboard

### CODE REVIEW

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on a6a6e3b69ceffb56f59caddd165f2166d5569b05<br /><br />component functions. trashed objects from all com_ tables. Used component to delete trashed items across the entire site.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27429">issues.joomla.org/tracker/joomla-cms/27429</a>.</sub>

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on 96beb04985940cdbfe7f29ef3c6f43d37121ef76<br /><br />Everything functions as expected after being added to list as core extension.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27429">issues.joomla.org/tracker/joomla-cms/27429</a>.</sub>

File: administrator/components/com_trash/trash.xml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+			<folder>Controller</folder>
+			<folder>forms</folder>
+			<folder>Model</folder>
+			<folder>services</folder>
+			<folder>tmpl</folder>
+			<folder>View</folder>
+		</files>
+		<languages folder="admin">
+			<language tag="en-GB">language/en-GB.com_trash.ini</language>
+			<language tag="en-GB">language/en-GB.com_trash.sys.ini</language>
```
Comment (infograf768) [inline administrator/components/com_trash/trash.xml]: I guess language prefix should be deleted.
See https://github.com/joomla/joomla-cms/pull/27430

**Comment by SharkyKZ** [comment]
I'm not sure if `com_checkin` is a good example. At least if we ever intend to support tables with custom column names. We have some ways of mapping column names to "standard" column names (`Joomla\CMS\Table\Table::setColumnAlias()` method or the UCM field mappings) but none of that works if we just run through all database tables.

File: administrator/components/com_trash/trash.xml (hunk 1 above)
Comment (brianteeman) [reply administrator/components/com_trash/trash.xml]: done

**Comment by brianteeman** [comment]
@SharkyKZ I would suggest that **if** that ever happens then this code can be updated.

File: administrator/language/en-GB/com_trash.ini [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+COM_TRASH="Empty Trash"
+COM_TRASH_CONFIGURATION="Trash: Options"
+COM_TRASH_DATABASE_TABLE="Database Table"
+COM_TRASH_DATABASE_TABLE_ASC="Database Table ascending"
+COM_TRASH_DATABASE_TABLE_DESC="Database Table descending"
+COM_TRASH_FILTER_SEARCH_DESC="Search database tables with items to delete."
+COM_TRASH_FILTER_SEARCH_LABEL="Search Database Tables"
+COM_TRASH_GLOBAL_DELETE="Maintenance: Global Empty Trash"
+COM_TRASH_ITEMS_TO_DELETE="Items to delete"
```
Comment (Quy) [inline administrator/language/en-GB/com_trash.ini]:
```suggestion
COM_TRASH_ITEMS_TO_DELETE="Items to Delete"
```

Comment (brianteeman) [reply administrator/language/en-GB/com_trash.ini]: I am accepting this - note that in com_checkin it needs to be changes as well

File: administrator/components/com_trash/Controller/DisplayController.php [hunk 3]
```diff
@@ -0,0 +1,122 @@
+<?php
+/**
+ * @package     Joomla.Administrator
+ * @subpackage  com_trash
+ *
+ * @copyright   Copyright (C) 2005 - 2019 Open Source Matters, Inc. All rights reserved.
```
Comment (ot2sen) [inline administrator/components/com_trash/Controller/DisplayController.php:6]: 2019-2020

File: administrator/components/com_trash/Model/TrashModel.php [hunk 4]
```diff
@@ -0,0 +1,270 @@
+<?php
+/**
+ * @package     Joomla.Administrator
+ * @subpackage  com_trash
+ *
+ * @copyright   Copyright (C) 2005 - 2019 Open Source Matters, Inc. All rights reserved.
```
Comment (ot2sen) [inline administrator/components/com_trash/Model/TrashModel.php:6]: 2019->2020

File: administrator/components/com_trash/View/Trash/HtmlView.php [hunk 5]
```diff
@@ -0,0 +1,113 @@
+<?php
+/**
+ * @package     Joomla.Administrator
+ * @subpackage  com_trash
+ *
+ * @copyright   Copyright (C) 2005 - 2019 Open Source Matters, Inc. All rights reserved.
```
Comment (ot2sen) [inline administrator/components/com_trash/View/Trash/HtmlView.php:6]: 2019->2020

File: administrator/components/com_trash/services/provider.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
+<?php
+/**
+ * @package     Joomla.Administrator
+ * @subpackage  com_trash
+ *
+ * @copyright   Copyright (C) 2005 - 2019 Open Source Matters, Inc. All rights reserved.
```
Comment (ot2sen) [inline administrator/components/com_trash/services/provider.php:6]: 2019->2020

File: administrator/components/com_trash/tmpl/trash/default.php [hunk 7]
```diff
@@ -0,0 +1,76 @@
+<?php
+/**
+ * @package     Joomla.Administrator
+ * @subpackage  com_trash
+ *
+ * @copyright   Copyright (C) 2005 - 2019 Open Source Matters, Inc. All rights reserved.
```
Comment (ot2sen) [inline administrator/components/com_trash/tmpl/trash/default.php:6]: 2019->2020

File: administrator/language/en-GB/com_trash.ini [hunk 8]
```diff
@@ -0,0 +1,28 @@
+; Joomla! Project
+; Copyright (C) 2005 - 2019 Open Source Matters. All rights reserved.
```
Comment (ot2sen) [inline administrator/language/en-GB/com_trash.ini:2]: 2019->2020

File: administrator/language/en-GB/com_trash.sys.ini [hunk 9]
```diff
@@ -0,0 +1,10 @@
+; Joomla! Project
+; Copyright (C) 2005 - 2019 Open Source Matters. All rights reserved.
```
Comment (ot2sen) [inline administrator/language/en-GB/com_trash.sys.ini:2]: 2019->2020

**Review by ot2sen — COMMENTED** [review]
just some update of year

**Comment by brianteeman** [comment]
@ot2sen no need to update the year. The build scripts will do that

File: administrator/components/com_trash/Model/TrashModel.php [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+			{
+				$column = 'state';
+			}
+			elseif (isset($fields['published']))
+			{
+				$column = 'published';
+			}
+
+			$query = $db->getQuery(true)
+					->delete($db->quoteName($tn))
```
Comment (richard67) [inline administrator/components/com_trash/Model/TrashModel.php]:
```suggestion
				->delete($db->quoteName($tn))
```

File: administrator/components/com_trash/Model/TrashModel.php [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+					->where( $column . ' = -2');
```
Comment (richard67) [inline administrator/components/com_trash/Model/TrashModel.php]:
```suggestion
				->where($column . ' = -2');
```

File: administrator/components/com_trash/Model/TrashModel.php [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+
+				if ($this->getState('filter.search') && stripos($tn, $this->getState('filter.search')) === false)
+				{
+					unset($tables[$i]);
+					continue;
+				}
+
+				$fields = $db->getTableColumns($tn);
+
+				// only work with the tables that have a state or published column
```
Comment (richard67) [inline administrator/components/com_trash/Model/TrashModel.php]:
```suggestion
				// Only work with the tables that have a state or published column
```

File: administrator/components/com_trash/Model/TrashModel.php [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+				if (!(isset($fields['state'])) && !(isset($fields['published'])))
+				{
+					unset($tables[$i]);
+					continue;
+				}
+
```
Comment (richard67) [inline administrator/components/com_trash/Model/TrashModel.php]:
```suggestion
```

File: administrator/components/com_trash/Model/TrashModel.php [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+				}
+				elseif (isset($fields['published']))
+				{
+					$column = 'published';
+				}
+
+				$query = $db->getQuery(true)
+					->select('COUNT(*)')
+					->from($db->quoteName($tn))
+					->where( $column . ' = -2');
```
Comment (richard67) [inline administrator/components/com_trash/Model/TrashModel.php]:
```suggestion
					->where($column . ' = -2');
```

**Review by richard67 — COMMENTED** [review]
PHPCS as reported by Drone

**Comment by richard67** [comment]
Great new feature ... I think it HAS to go into 4.0 ... so I hope it will be ready before beta.

@brianteeman Is it still draft/WiP? Or is it (beside PHPCS) ready for test? Let us know please when we shall test.

**Comment by brianteeman** [comment]
it is a draft - as in there are bits not done such as install sql AND the questions in the original post. I am confident that the code works just dont know if it is something people think is a good idea etc

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on 292e43e7e236c4fa7a810ee458f40b0ed2ea031b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27429">issues.joomla.org/tracker/joomla-cms/27429</a>.</sub>

**Comment by jwaisner** [comment]
@richard67 Can you please give this one a test?

@brianteeman I think it is a good idea. Now how to handle the install SQL is beyond me.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27429">issues.joomla.org/tracker/joomla-cms/27429</a>.</sub>

**Comment by richard67** [comment]
@jwaisner As far as I understood @brianteeman this PR is still a draft and not ready. And it is late now here in my time zone and I have to work tomorrow, so right I won't test for sure.

**Comment by HLeithner** [comment]
The idea of a global trash is good, but using com_checkin as base is not the best approach.  Sharky already mentioned the this component loads all tables and tries to figure out if there is a trash able item or not. 

The right way would be that each component has to register for the global trash.

**Comment by brianteeman** [comment]
> The right way would be that each component has to register for the global trash.

There is perfect engineering and there is working engineering

**Comment by SharkyKZ** [comment]
Well, this isn't working. Items should be deleted through the model to make sure all associated data is deleted and proper plugin events are triggered.

**Comment by brianteeman** [comment]
Can you please provide a specific example of where its not working

**Comment by SharkyKZ** [comment]
Everywhere. None of associated data is deleted because none of the code meant for handling item deletion is being executed this way. Examples of remaining loose data include Asset record (in `#__assets` table), Language Associations, Custom Fields data, Tag mappings, Content History and 3rd party plugin data.

**Comment by brianteeman** [comment]
forgive me for trying to be constructive and add a requested feature

**Comment by SharkyKZ** [comment]
You don't need to apologize. It's a good feature but it needs a different approach. I, for one, don't know what that approach would be 😕 .

**Comment by brianteeman** [comment]
re-oponed in draft status as a reminder to myself to work on this

**Comment by brianteeman** [comment]
Closed - no interest

**Pull request closed without merging**